Feature/dashboard latest news/watch

Udit-takkar opened the pull request:
## Overview

**What features/fixes does this PR include?**
Dashboard latest news Watch section.

## Implementation
used onScreen hook to implement infinite scrolling and google data api to get yt videos fro MetaMedia YT Channel.

## Asset
https://user-images.githubusercontent.com/53316345/143764311-80c84136-d652-4c88-aed9-ac4f66439f51.mov

Udit-takkar commented:
Currently i have put the google api key in my .env so this feature would throw error on vercel deployed.

Udit-takkar commented:
PS: Only my last two commits implement this PR feature

Udit-takkar commented:
Currently testing if `useSwrInfinite`  can be used here.

vidvidvid commented:
> PS: Only my last two commits implement this PR feature

Wait until https://github.com/MetaFam/TheGame/pull/900 gets merged into develop, then rebase your branch onto develop and you should be fine.

vidvidvid commented:
I get an error when opening the watch tab:

```
Failed to load resource: the server responded with a status of 400 ()

dashboard-499bc353e49073161178.js:1 
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'totalResults')
    at /_next/static/chunks/pages/dashboard-499bc353e49073161178.js:1
    at l (/_next/static/chunks/main-444f4a448dcaee7f542b.js:1)
    at Generator._invoke (/_next/static/chunks/main-444f4a448dcaee7f542b.js:1)
    at Generator.next (/_next/static/chunks/main-444f4a448dcaee7f542b.js:1)
    at n (/_next/static/chunks/pages/_app-30de55cd029ba307311f.js:1)
    at s (/_next/static/chunks/pages/_app-30de55cd029ba307311f.js:1)
```

Udit-takkar commented:
> I get an error when opening the watch tab:
> 
> ```
> Failed to load resource: the server responded with a status of 400 ()
> 
> dashboard-499bc353e49073161178.js:1 
> Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'totalResults')
>     at /_next/static/chunks/pages/dashboard-499bc353e49073161178.js:1
>     at l (/_next/static/chunks/main-444f4a448dcaee7f542b.js:1)
>     at Generator._invoke (/_next/static/chunks/main-444f4a448dcaee7f542b.js:1)
>     at Generator.next (/_next/static/chunks/main-444f4a448dcaee7f542b.js:1)
>     at n (/_next/static/chunks/pages/_app-30de55cd029ba307311f.js:1)
>     at s (/_next/static/chunks/pages/_app-30de55cd029ba307311f.js:1)
> ```

Yeah that's what i have mentioned in the comment above. Wait I'll send you the google api key on DM

Udit-takkar commented:
> Wait until #900 gets merged into develop, then rebase your branch onto develop and you should be fine.
Yes
 
> Also could you elaborate on how would you like to use `useSwrInfinite` and how is it different from `useSwr`?
useSwrInfinite is for sending paginated queries for infinite scrolling

vidvidvid commented:
> Yeah that's what i have mentioned in the comment above. Wait I'll send you the google api key on DM

Sorry, my bad - should have read the comments first.

vidvidvid commented on packages/web/components/Dashboard/LatestContentTabs/Watch.tsx:
```diff
@@ -1,8 +1,107 @@
+      const res = await fetch(URL).then((r) => r.json());
+
+      setNextPageToken(res.nextPageToken);
+      setVideos(res.items);
```
```suggestion
      setVideos(res.items.map(item => ({ etag: item.etag, id: item.id, kind: item.kind })));
```

vidvidvid replied:
actually you need different data than here. on discord you said you only need `etag`, `id` and `kind` if I remember corretly. But looking at the code:
```
 const url = `https://www.youtube.com/embed/${video?.snippet?.resourceId?.videoId}`;
        return (
          <AspectRatio key={video.id} ratio={2 / 1}>
            <Box
              as="iframe"
              title={video?.snippet?.title}
```
It seems like you need `video?.snippet?.resourceId?.videoId`, `video.id` and `video?.snippet?.title`. So if you don't need any other data to display the video, please map the items that you receive into objects that will contain only these three pieces of data:
```
Video: {
 videoId: '',
 id: '',
 title: '',
}
By the way, what is the difference between `videoId` and `id`?
```

vidvidvid commented on packages/web/components/Dashboard/LatestContentTabs/Watch.tsx:
```diff
@@ -1,8 +1,107 @@
+  }, [count]);
+
+  useEffect(() => {
+    const load = async () => {
```
Can you use the `useSwr` package?

vidvidvid commented on packages/web/components/Dashboard/LatestContentTabs/Watch.tsx:
```diff
@@ -1,8 +1,107 @@
+      setIsFetchingMore(true);
+      const fetchMoreURL = `${URL}&pageToken=${nextPageToken}`;
+
+      fetch(fetchMoreURL)
```
can you refactor that to use the `useSwr` package?

Udit-takkar replied:
ok i will have refactor in `useSwrInfinite` but i am not sure if i should do it now or after we build component that would display read,watch and listen combined.

vidvidvid reviewed (changes requested):
Overall smooth solution and nice job, just a syntactic improvement suggestion (which we discussed in discord) and a performance improvement ([which you told me about first :P](https://github.com/MetaFam/TheGame/pull/927#issuecomment-974855189))

vidvidvid commented on packages/web/components/Dashboard/LatestContentTabs/Watch.tsx:
```diff
@@ -1,8 +1,107 @@
+  ) : (
+    <Box>
+      {videos.map((video: Video) => {
+        const url = `https://www.youtube.com/embed/${video?.snippet?.resourceId?.videoId}`;
```
`video?.`
here asking if video exists is not needed, since you're already looping through an array of videos, that have to exist.

Udit-takkar commented on packages/web/components/Dashboard/LatestContentTabs/Watch.tsx:
```diff
@@ -1,8 +1,107 @@
+  width: number;
+  height: number;
+};
+interface Video {
+  etag: string;
+  id: string;
+  kind: string;
+  snippet: {
+    channelId: string;
+    channelTitle: string;
+    description: string;
+    playlistId: string;
+    position: number;
+    publishedAt: string;
+    resourceId: {
+      kind: string;
+      videoId: string;
+    };
+    title: string;
+    videoOwnerChannelId: string;
+    videoOwnerChannelTitle: string;
+    thumbnails: {
+      default?: Thumbnail;
+      high?: Thumbnail;
+      medium?: Thumbnail;
+      standard?: Thumbnail;
+      maxres?: Thumbnail;
+    };
+  };
+}
```
actually i tried using the map on `res.items` but the current element while mapping was  required to give a type which would be Video.

vidvidvid replied:
what do you mean with 'current element'?

vidvidvid commented on packages/web/components/Dashboard/LatestContentTabs/Watch.tsx:
```diff
@@ -1,8 +1,107 @@
+  width: number;
+  height: number;
+};
+interface Video {
+  etag: string;
+  id: string;
+  kind: string;
+  snippet: {
+    channelId: string;
+    channelTitle: string;
+    description: string;
+    playlistId: string;
+    position: number;
+    publishedAt: string;
+    resourceId: {
+      kind: string;
+      videoId: string;
+    };
+    title: string;
+    videoOwnerChannelId: string;
+    videoOwnerChannelTitle: string;
+    thumbnails: {
+      default?: Thumbnail;
+      high?: Thumbnail;
+      medium?: Thumbnail;
+      standard?: Thumbnail;
+      maxres?: Thumbnail;
+    };
+  };
+}
```
```suggestion
interface Video {
  id: string;
  videoId: string;
  title: string;
}
```

vidvidvid commented on packages/web/components/Dashboard/LatestContentTabs/Watch.tsx:
```diff
@@ -1,8 +1,107 @@
+      const res = await fetch(URL).then((r) => r.json());
+
+      setNextPageToken(res.nextPageToken);
+      setVideos(res.items);
```
```suggestion
      setVideos(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        res.items.map((video: any) => ({
          id: video.id,
          videoId: video.snippet.resourceId.videoId,
          title: video.snippet.title,
        })),
      );
```

Udit-takkar replied:
Yeah this is what i was talking about this. So should i just use any ?

vidvidvid replied:
in this case i see no harm in it. it enables you to write less and more concise.

Udit-takkar replied:
Ok

vidvidvid commented on packages/web/components/Dashboard/LatestContentTabs/Watch.tsx:
```diff
@@ -1,8 +1,107 @@
+    <LoadingState />
+  ) : (
+    <Box>
+      {videos.map((video: Video) => {
+        const url = `https://www.youtube.com/embed/${video?.snippet?.resourceId?.videoId}`;
+        return (
+          <AspectRatio key={video.id} ratio={2 / 1}>
+            <Box
+              as="iframe"
+              title={video?.snippet?.title}
+              src={url}
+              allowFullScreen
+              pt={4}
+            />
```
```suggestion
      {videos.map(({ videoId, id, title }: Video) => {
        const url = `https://www.youtube.com/embed/${videoId}`;
        return (
          <AspectRatio key={id} ratio={2 / 1}>
            <Box as="iframe" title={title} src={url} allowFullScreen pt={4} />
```

alalonde commented on packages/web/config.ts:
```diff
@@ -30,4 +30,5 @@ export const CONFIG = {
     'https://d12-a-ceramic.3boxlabs.com',
   actionsURL: process.env.NEXT_ACTIONS_URL || 'http://localhost:4000',
   calendarId: 'nih59ktgafmm64ed4qk6ue8vv4@group.calendar.google.com',
+  googleDataApiKey: process.env.NEXT_PUBLIC_GOOGLE_API_KEY,
```
```suggestion
  googleDataApiKey: process.env.NEXT_PUBLIC_YOUTUBE_API_KEY,
```

Since this is scoped just for the YouTube Data API

Udit-takkar replied:
ok

Udit-takkar commented:
I would add `useSwrInfinite` once we make the component in which everything would display together.

Udit-takkar commented:
@alalonde  can you review this?

alalonde commented:
I added the new key (that @Udit-takkar DM'ed me) to vercel `NEXT_PUBLIC_YOUTUBE_API_KEY` but the preview still can't load the youtube content... any ideas?

https://the-game-in1p81yke-metafam.vercel.app/dashboard

Udit-takkar commented:
> I added the new key (that @Udit-takkar DM'ed me) to vercel `NEXT_PUBLIC_YOUTUBE_API_KEY` but the preview still can't load the youtube content... any ideas?
> 

the value of the key is passed undefined. I had also created `.env.local` and added the env variable there and it was loading correctly from there.
or maybe we can add it in `next.config.js`
```
module.exports = {
  reactStrictMode: true,
  env: {
    BASE_URL: process.env.BASE_URL,
  }
}
```

alalonde commented:
Triggering a redeploy..

Udit-takkar commented:
@alalonde  it is ready
